### PR TITLE
Refactor optimizer test for multi-backend

### LINKED ISSUE #7584: `OptimizerHook` ChainerX support
Part of #7541

### PR DESCRIPTION
**Pull request by niboshi**
~`UpdateRule.update` does not handle multiple devices with loss scaling.~

This PR includes test refactoring. (part of #7541)

Fixes #7584

### CODE REVIEW

**Comment by hvy** [comment]
Which tests are testing the change introduced in the `UpdateRule`? Is it worth splitting the PR since the test refactoring is quite large and most seem unrelated to the change in the `UpdateRule`?

**Comment by niboshi** [comment]
Separated the loss-scale fix as #7590. That PR depends on this PR.

File: chainer/optimizer.py [hunk 1]
```diff
@@ -192,6 +192,10 @@ def update(self, param):
 
         self.t += 1
 
+        with chainer.using_device(param.device):
```
[the remark this answers is not in the document]
Comment (niboshi) [reply chainer/optimizer.py]: This fix was separated as #7594.

File: tests/chainer_tests/test_optimizer.py [hunk 2]
```diff
@@ -638,79 +629,63 @@ def setUp(self):
             lr = self.loss_scale
             for i in range(2):
                 self.target[i].param._loss_scale = self.loss_scale
```
Comment (hvy) [inline tests/chainer_tests/test_optimizer.py]: Doesn't have to be addressed in this PR but there must be a better way to test the loss scale other than modifying a private attribute 🤔

File: tests/chainer_tests/test_optimizer.py [hunk 3]
```diff
@@ -638,79 +629,63 @@ def setUp(self):
...
-        else:
-            assert orig_xp is np
-            self.target.to_device('native:0')
-        self.optimizer.setup(self.target)
-
-    def check_update(self):
-        self.optimizer.update()
-        xp = backend.get_array_module(self.target[0].param)
+    def test_update(self, backend_config):
+        if backend_config.device == chainer.get_device('@cupy:1'):
```
Comment (hvy) [inline tests/chainer_tests/test_optimizer.py]: Seems like this test is executed on Travis.

File: tests/chainer_tests/test_optimizer.py [hunk 4]
```diff
@@ -15,6 +15,23 @@
 from chainer.testing import attr
 import chainerx
 
+if chainerx.is_available():
+    import chainerx.testing
```
[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: It's used via `device.xp.testing`.

**Comment by hvy** [comment]
Wrote some minor comments. Could you take a look? Travis also seems to be failing due to a test depending on CuPy being run.

File: tests/chainer_tests/test_optimizer.py (hunk 2 above)
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: Possibly. Indeed it's something better discussed at another place.

File: tests/chainer_tests/test_optimizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/test_optimizer.py]: Ah, that's not really obvious. Could you leave a comment?

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: I'm not sure we should leave such comment.
If we use `chainerx.testing....` somewhere else, that comment will be obsolete and it's difficult to notice you should remove it.

File: tests/chainer_tests/test_optimizer.py (hunk 3 above)
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: Thanks. Fixed.

File: tests/chainer_tests/test_optimizer.py [hunk 5]
```diff
@@ -74,96 +91,127 @@ def __call__(self, key, value):
         return value
 
 
-class TestUpdateRule(unittest.TestCase):
-
+def _create_update_rule(has_states=False):
```
[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: Fixed.

File: tests/chainer_tests/test_optimizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/test_optimizer.py]: Thanks for your comment, but Is it not more important to be informative to the general maintainer? It's difficult to notice if the comment is obsolete and remove it when it becomes relevant and I think that someone that'll use `chainerx.testing` in this file is likely to stumble upon this comment anyway. 

This is really not a strong opinion as it's about a comment in a test file but just my thought.

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: >I think that someone that'll use `chainerx.testing` in this file is likely to stumble upon this comment anyway.

As stated above, I doubt that. I would just write `chainerx.testing....` and never mind the imports.

[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/test_optimizer.py]: Hm, you disagree on the other points in the comment above as well?

[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/test_optimizer.py]: Hm, you disagree on the other points in the comment above as well?

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: >Is it not more important to be informative to the general maintainer?

Do you mean this one?
I don't think the comment would be that helpful.
For example, if you see `from chainer.testing import attr`, are you going to check where `attr` is actually used?

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/test_optimizer.py]: >Is it not more important to be informative to the general maintainer?

Do you mean this one?
I don't think the comment would be that helpful.
For example, if you see `from chainer.testing import attr`, are you going to check where `attr` is actually used?
Even if it's helpful in some degree, that won't justify the downside mentioned above.

[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/test_optimizer.py]: Okay, I see your point. Let's leave it as is.

**Comment by hvy** [comment]
Jenkins, test this please.

**Pull request merged**